JENKINS-57484 - Admin's Initial API token is documented. (#4022)
jenkins-infra/jenkins.io: content/doc/book/managing/system-properties.adoc

    Development-mode only: Set to `true` to not skip showing the setup wizard during Jenkins development.
    This property is only effective the first time you run Jenkins in given `JENKINS_HOME`.

- name: jenkins.install.SetupWizard.adminInitialApiToken
  tags:
  - security
  - packaging

>   def: undefined

Reviewer: Wouldn't it be clearer to the reader if the last sentence were included as the explanation of the default behavior?  That would make the default value clear to readers and follow the same pattern used by [executable-war](https://www.jenkins.io/doc/book/managing/system-properties/#executable-war).
```suggestion
  def: The default admin account will not have an API Token unless a value is provided for this system property
```

  since: 2.260
  description: |
    This property determines the behavior during the SetupWizard install phase concerning the API Token creation for the initial admin account.
    The behavior depends on the provided value:

>     `true`:: A token is generated using random value at startup and the information is put in the file `$JENKINS_HOME/secrets/initialAdminApiToken`. This approach is the more interesting (+ secure) when you have read access to the Jenkins home directory.

Reviewer 2: sentence doesn't quite make sense, I think the best fix is to remove it
```suggestion
    `true`:: A token is generated using random value at startup and the information is put in the file `$JENKINS_HOME/secrets/initialAdminApiToken`.
```

    "token" in plain text:: A fixed API Token will be created for the user with provided value as the token.

>     "@[file-location]" which contains plain text value of the token:: A fixed API Token will be created for the user with the value read from the file. The application will not delete the file after read, so the script is responsible to clean up the stuff

Reviewer 2: ```suggestion
    "@[file-location]" which contains plain text value of the token:: A fixed API Token will be created for the user with the value read from the file. Jenkins will not delete the file after read, so the script is responsible to remove it when no longer needed.
```

>     Token format is `[2-char hash version][32-hex-char of secret]`, where the hash version is currently only 11. E.g. 110123456789abcdef0123456789abcdef.

Reviewer 2: it would be good to link to a script / method of generating this, especially to prevent people from using the example one provided

Author: @Wadeck can you help with this? In my test, I indeed just copy-pasted this example.

Reviewer 3: Possibilities:
- through online tool: https://www.browserling.com/tools/random-hex, How many digit = 32, and then prepend 11 for the version
- in shell: ` echo "11$(openssl rand -hex 16)"`
- in JavaScript (inside your browser) like
```
const genRanHex = size => [...Array(size)].map(() => Math.floor(Math.random() * 16).toString(16)).join('');
console.log('11' + genRanHex(32));
```
- in Java like
```
public static void main(String[] args) throws NoSuchAlgorithmException {
    Random secure = SecureRandom.getInstanceStrong();

    byte[] bytes = new byte[16];
    secure.nextBytes(bytes);
    StringBuilder result = new StringBuilder("11");
    for (byte aByte : bytes) {
        result.append(String.format("%02x", aByte));
    }
    System.out.println(result.toString());
}
```

Tip: the need for a secure random generator is less important as it is inside an application as you are generating this token out of the execution time and thus, to determine what was generated, the attacker is required to have access to your machine (or you are using a pretty dumb generator)

>     When the API Token is generated using this system property, it's strongly recommended that you are revoking it during your installation script using the other ways at your disposal so that you have a fresh (random) token with less traces for your script. See https://javadoc.jenkins.io/jenkins/security/ApiTokenProperty.html#generateNewToken-java.lang.String-[ApiTokenProperty#generateNewToken(String)] and https://javadoc.jenkins.io/jenkins/security/ApiTokenProperty.html#revokeAllTokensExceptOne-java.lang.String-[ApiTokenProperty#revokeAllTokensExceptOne(String)] for scripting methods or using the web API calls:

Reviewer: I'm not understanding the recommendation to revoke the API token during installation.  Is it recommended that a single API token is all that is needed or is it recommended that most of the time there is no API token defined?
```suggestion
    When the API Token is generated using this system property, it should be revoked during the installation script using the other ways at your disposal so that you have a fresh (random) token with less traces for your script. See https://javadoc.jenkins.io/jenkins/security/ApiTokenProperty.html#generateNewToken-java.lang.String-[ApiTokenProperty#generateNewToken(String)] and https://javadoc.jenkins.io/jenkins/security/ApiTokenProperty.html#revokeAllTokensExceptOne-java.lang.String-[ApiTokenProperty#revokeAllTokensExceptOne(String)] for scripting methods or using the web API calls:
```
